rest-service/manager_rest/rest/resources_v3_1/executions.py

```diff
@@ -13,12 +13,122 @@
+        if args['status']:
+            if args['status'] not in ExecutionState.END_STATES:
+                raise ValueError(
+                    'Can\'t filter by execution status `{0}`. '
+                    'Allowed statuses are: {1}'.format(
+                        args['status'], ExecutionState.END_STATES)
+                )
+            filters['status'] = args['status']
+        else:
+            filters['status'] = ExecutionState.END_STATES
+
+        executions = sm.list(models.Execution,
+                             filters=filters,
+                             all_tenants=args['all_tenants'])
+        dep_creation_execs = {}
+        for execution in executions:
+            if execution.workflow_id == 'create_deployment_environment' and \
+                    execution.status == 'terminated':
+                dep_creation_execs[execution.deployment_id] = \
+                    dep_creation_execs.get(execution.deployment_id, 0) + 1
+
+        executions_to_delete = []
+
+        if request_dict.get('keep_days'):
+            now_utc = datetime.utcnow()
+            requested_time = (datetime(*now_utc.timetuple()[:3])
+                              - timedelta(days=request_dict['keep_days']-1))
+            for execution in executions:
+                creation_time = datetime.strptime(execution.created_at,
+                                                  '%Y-%m-%dT%H:%M:%S.%fZ')
+                if creation_time < requested_time and \
+                        self._can_delete_execution(execution,
+                                                   dep_creation_execs):
+                    executions_to_delete.append(execution)
+                    sm.delete(execution)
+        elif keep_since:  # UNIX date format: %Y-%m-%d
+            for execution in executions:
+                creation_time = datetime.strptime(execution.created_at,
+                                                  '%Y-%m-%dT%H:%M:%S.%fZ')
+                if creation_time < keep_since and \
+                        self._can_delete_execution(execution,
+                                                   dep_creation_execs):
+                    executions_to_delete.append(execution)
+                    sm.delete(execution)
+        elif request_dict.get('keep_last'):
+            num_to_delete = len(executions) - request_dict['keep_last']
+            for execution in executions:
+                if self._can_delete_execution(execution, dep_creation_execs):
+                    executions_to_delete.append(execution)
+                    sm.delete(execution)
+                    num_to_delete -= 1
+                if num_to_delete == 0:
+                    break
+        else:
+            raise ValueError('Must specify one of the options keep_last | '
+                             'keep_days | keep_since_date')
+        return ListResult(executions_to_delete, {'pagination': pagination})
```

Comment:
returning the whole list of executions seems wasteful, that might well be many megabytes, and it seems we only do it so that the cli can display the count (!) of deleted executions
for delete, i'd say it's fine to just return nothing instead. Maybe the count in the pagination.